File: backend/utils/hackerrank.py
"""
utils/hackerrank.py
Fetches user statistics from HackerRank using their internal REST API.
"""
import requests

HACKERRANK_PROFILE_URL = "https://www.hackerrank.com/rest/contests/master/users/{username}/profile"

def get_hackerrank_stats(username: str) -> dict | None:
    """
    Fetch user's total score and rank from HackerRank.
    """
    url = HACKERRANK_PROFILE_URL.format(username=username)
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36",
    }
    
    try:
        resp = requests.get(url, headers=headers, timeout=10)
        
        if resp.status_code != 200:
            print(f"HackerRank API returned status {resp.status_code}")
            return None
            
        data = resp.json()
        model = data.get("model", {})
        
        if not model:
            return None
            
        # HackerRank has multiple scores for different tracks. 
        # We'll try to aggregate or pick the primary ones.
        # For simplicity, let's take the 'total_count' or specific badges if available.
        # However, the profile REST API gives 'score' and 'rank' in some fields.
        
        return {
            "hr_score": int(model.get("total_score", 0)),
            "hr_rank": model.get("rank", 0),
        }
        
    except Exception as e:
        print(f"Error fetching HackerRank stats for {username}: {e}")
        return None

def sync_hackerrank_stats(username: str) -> dict:
    """
    Returns a dict ready to update the Profile model.
    """
    stats = get_hackerrank_stats(username)
    return stats if stats else {}

File: backend/src/utils/hackerrank.js
/**
 * utils/hackerrank.js
 * Wrapper to fetch user statistics from HackerRank using public REST endpoints.
 */

import axios from 'axios';

const HACKERRANK_BASE = 'https://www.hackerrank.com/rest/hackers';

export async function getHackerrankStats(username) {
  if (!username || typeof username !== 'string') return null;
  const cleanUsername = username.trim();
  if (!cleanUsername) return null;

  try {
    const profileResp = await axios.get(`${HACKERRANK_BASE}/${encodeURIComponent(cleanUsername)}/profile`, {
      timeout: 8000,
      headers: {
        'User-Agent':
          'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
      },
    });

    if (profileResp.status !== 200 || !profileResp.data?.model) {
      return null;
    }

    const model = profileResp.data.model;

    // Fetch badges
    let badgesCount = 0;
    let totalScore = 0;
    let problemsSolved = 0;

    try {
      const badgesResp = await axios.get(`${HACKERRANK_BASE}/${encodeURIComponent(cleanUsername)}/badges`, {
        timeout: 8000,
        headers: {
          'User-Agent':
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
        },
      });

      if (badgesResp.status === 200 && Array.isArray(badgesResp.data?.models)) {
        const badges = badgesResp.data.models;
        badgesCount = badges.length;
        for (const badge of badges) {
          totalScore += parseInt(badge.hacker_score || 0, 10);
          problemsSolved += parseInt(badge.solved || 0, 10);
        }
      }
    } catch {
      // Badges endpoint optional
    }

    return {
      hr_badges: badgesCount,
      hr_score: totalScore || parseInt(model.score || 0, 10),
      hr_problems_solved: problemsSolved || parseInt(model.solved_challenges || 0, 10),
    };
  } catch (err) {
    console.log(`[HACKERRANK] Error fetching stats for ${cleanUsername}: ${err.message}`);
    return null;
  }
}

export async function syncHackerrankStats(username) {
  const stats = await getHackerrankStats(username);
  return stats || {};
}
